### PR TITLE
Install workflow

### PR DESCRIPTION
**Pull request by mvdbeek**
This adds the possibility to import workflows.
The python script is based on @cameron-jack's [PR](https://github.com/bgruening/docker-galaxy-stable/pull/137) to docker-galaxy.

### CODE REVIEW

**Comment by mvdbeek** [comment]
Unfortunately Galaxy doesn't expose workflow sharing via the API ... that would be a nice addition.

**Comment by bgruening** [comment]
Thanks @mvdbeek! My feeling is that we should have a smarter script here, that also ensures (if possible) that all dependencies are installed that are needed for a workflow. I guess this is what I would expect from this role.

**Comment by mvdbeek** [comment]
Hmm, that's not easy, since the workfow steps don't have a name/owner/toolshed/changeset id combination that we would need (I would love to have this implemented in whatever format will succeed *.ga !!!). Also, which tool panel section/id would we choose?

**Comment by mvdbeek** [comment]
I'm not sure how tolerant the .ga import is, but if additional fields in a workflow file don't break the import, I could perhaps craft a PR to add this to core Galaxy ... could be worth a try, I guess.

**Comment by bgruening** [comment]
I did not said it's easy :)
I think the id is unique enough and should make it possible to get the correct repository: `"tool_id": "toolshed.g2.bx.psu.edu/repos/devteam/ncbi_blast_plus/ncbi_tblastn_wrapper/0.1.01"`

The panel section is a problem we face since years. I'm voting for default categories that can be tagged in the tool. But again work on Galaxy is needed here.

**Comment by mvdbeek** [comment]
@bgruening 
I already tried this, there's no reliable way to get the version aligned to a changeset revision, and the tool_id structure changes with the structure of the repo.

```
"tool_id": "toolshed.g2.bx.psu.edu/repos/devteam/ncbi_blast_plus/ncbi_tblastn_wrapper/0.1.01"
```

would be be

```
"tool_id": "toolshed.g2.bx.psu.edu/repos/devteam/ncbi_tblastn_wrapper/0.1.01"
```

if it was not a suite of tools ...
But I'm opening an issue on galaxy to get feedback on that issue

**Comment by bgruening** [comment]
Mh, the toolshed has a tool-view that lists all tool-ids, with associated repositories. But I guess there is no API which we can reuse.

**Comment by mvdbeek** [comment]
It seems that you can add fields to .ga files without breaking compatibility, so I would prefer to collect all tools and put them in a new requirements field.

**Comment by bgruening** [comment]
Sure, this would be the best solution. @jmchilton is working on a version 2.0 of this workflow definition as I know so maybe he has more input.

**Comment by jmchilton** [comment]
Open a Galaxy PR and toss more tool shed fields into the workflow export - that should be very easy. I'll give you the method... https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/workflows.py#L488.

Maybe inside of 

```
if module.type == 'tool':
```

add the properties to `step_dict`. `module.tool` should be defined and should have all the fields you need (hopefully).

`.ga` will be supported indefinitely and should not be considered deprecated in anyway. I've been playing with sane workflow definitions - but they are for humans to write not for Galaxy to export at this time.

**Comment by mvdbeek** [comment]
> add the properties to step_dict. module.tool should be defined and should have all the fields you need (hopefully).

I was thinking about collecting the tool information and putting it in a new requirements field at the top of the workflow, so that this information is not repeated when the same tool is used many times. Thanks, that looks pretty doable!

**Comment by jmchilton** [comment]
@mvdbeek It seems more precise to put it next to the actual tool, so I'd prefer to see it at least there. Downstream tools can extract this out from the steps pretty easily I think, if you want to duplicate it at the top level though I won't object.

File: defaults/main.yml [hunk 1]
```diff
@@ -23,6 +23,10 @@ galaxy_tools_base_dir: /tmp
 # Blank variable to make sure it's defined
 galaxy_tools_api_key: ''
 
+# List of workflow files to be installed
+
+galaxy_tools_workflows: []
```
Comment (afgane) [inline defaults/main.yml:31]: I think it'd be good to also include a sample of the values that would be used to populate this list.

Comment (mvdbeek) [inline defaults/main.yml:31]: Okay, I will add a sample workflow.

File: tasks/main.yml [hunk 2]
```diff
@@ -6,5 +6,8 @@
 - include: tools.yml
   when: galaxy_tools_install_tools
 
+- include: workflows.yml
+  when: galaxy_tools_workflows|length > 0
```
Comment (mvdbeek) [inline tasks/main.yml]: @afgane let me know if you prefer an explicit `galaxy_tools_install_workflow` variable.

Comment (afgane) [inline tasks/main.yml]: I'd say we should have that variable. This is shorter but having the variable is more consistent with the rest of the code and also if someone has workflows listed, they will not have to comment them out.

File: defaults/main.yml [hunk 3]
```diff
@@ -23,6 +23,11 @@ galaxy_tools_base_dir: /tmp
 # Blank variable to make sure it's defined
 galaxy_tools_api_key: ''
 
+# List of workflow files to be installed. Use empty list ([]) to skip installing workflows
+
+galaxy_tools_workflows:
+  - sample-workflow.ga
```
Comment (afgane) [inline defaults/main.yml]: How about switching this around and having the empty list be the default but including the example in the documentation line? 

Thanks for including the sample file.

**Comment by mvdbeek** [comment]
@bgruening @afgane 
Reflecting on the issue, I'm not a fan of automagically installing the dependencies that we can now infer from workflows in the workflow task. I think this defeats the purpose of declaratively managing servers with ansible.
I think we should instead have a check that parses the workflow and verifies that the required tools are installed and warns the user of missing tools, prompting them to be included into tool_list.yml (for which I can write a script, no problemo!).

**Comment by afgane** [comment]
@mvdbeek I tend to agree with your thinking. On one end, Galaxy does not automatically install any missing tools either. Two, with some modifications (i.e., not having the install script copied to the target machine), it would be possible for a user to run this script and batch-import their workflows as well. 
However, it would be kind of nice to also be able to do what Bjoern is saying and do a fire-and-forget thing where the tools get automagically installed with the workflows being imported (imagine setting up a temporary training server and simply importing a couple of workflows while the tools install with those). So, I guess an option that defaults to `False` to perform the tool install would be the best?

**Comment by afgane** [comment]
Testing this a bit, what do you guys think about checking for the workflow uuid before doing the import and not reimporting the same workflow? That way the script could be used in an idempotent fashion.

**Comment by cameron-jack** [comment]
Thank you for this idea. This could easily go into my python script that performs the upload. Would you like me to make that change? Should be a couple of lines to do the check.
Regards,
Cam

**Comment by bgruening** [comment]
> Reflecting on the issue, I'm not a fan of automagically installing the dependencies that we can now infer from workflows in the workflow task. I think this defeats the purpose of declaratively managing servers with ansible.

I don't see the point here. What is the difference than installing a workflow-with-tools or a simple tool.yaml file in regard of defeating ansible deployments?

KISS! Why making it more complex than needed? My initial complain was that this is a role for installing tools ... so offering a workflow copying without installing tools seems to be counter-intuitive.

I like @afgane idea of making it possible to disable tool installations. I mean no one will be forced to use this and everyone can simply include all requirements into a tool.yaml file.

**Comment by mvdbeek** [comment]
> I don't see the point here. What is the difference than installing a workflow-with-tools or a simple tool.yaml file in regard of defeating ansible deployments?

Well, if you put the tools in the tool.yaml file, you know exactly which tools at which version will get installed in which tool panel section. If you rely on let's say 20 or more different workflows defining the tools, you'll have to walk through each of them to check the revision and add them to the correct tool panel section. And when the workflows evolve you'll have to do the same thing again ...
Then again, coding-wise this is a minor detail, so I'll be happy to make it possible.

> My initial complain was that this is a role for installing tools ... so offering a workflow copying without installing tools seems to be counter-intuitive.

True, I see that. My preferred solution would be to just report which tools are missing. But if that's a config option I'm happy too.

@cameron-jack 

> Thank you for this idea. This could easily go into my python script that performs the upload. Would you like me to make that change? Should be a couple of lines to do the check.
> Regards,
> Cam

Go ahead!

@bgruening @afgane @cameron-jack 
Would you be opposed to merging this as is? I'll continue on working on the tool-install stuff then.
I prefer smaller PRs that evolve quickly, as @cameron-jack can add the uuid check based on the master branch.

**Comment by afgane** [comment]
Conceptually, it looked like we were all on the same page so it's now merged (clearly) - thanks for evolving this (and looking forward to the uuid enhancement).

**Comment by jmchilton** [comment]
My two cents is that forcing the tools in the yaml file is a more structured and ultimately better approach (control sections, etc...). If y'all do implement auto-installation of missing tools though, I'd encourage you to move this functionality upstream and add a flag to the Galaxy API to do this.

**Comment by mvdbeek** [comment]
> If y'all do implement auto-installation of missing tools though, I'd encourage you to move this functionality upstream and add a flag to the Galaxy API to do this.

Awesome, I was already thinking about this. Shouldn't be too hard (if we talk API only). On the UI (and API) side I guess there's also work to be done on managing tool panel sections.

**Pull request merged**